python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -93,44 +137,29 @@ def impl(ptr, offset):
     return impl
 
 
-class RawPointer:
+class RawPointer(IteratorBase):
     def __init__(self, ptr, ntype):
-        self.val = ctypes.c_void_p(ptr)
-        data_as_ntype_pp = numba.types.CPointer(numba.types.CPointer(ntype))
-        self.ntype = ntype
-        self.prefix = "pointer_" + ntype.name
-        self.ltoirs = [
-            _ncc(
-                f"{self.prefix}_advance",
-                RawPointer.pointer_advance,
-                numba.types.void(data_as_ntype_pp, _DISTANCE_NUMBA_TYPE),
-            ).ltoir,
-            _ncc(
-                f"{self.prefix}_dereference",
-                RawPointer.pointer_dereference,
-                (data_as_ntype_pp,),
-            ).ltoir
-        ]
-
-    # Exclusively for numba.cuda.compile (this is not an actual method).
-    def pointer_advance(this, distance):
-        this[0] = this[0] + distance
-
-    # Exclusively for numba.cuda.compile (this is not an actual method).
-    def pointer_dereference(this):
-        return this[0][0]
-
-    @property
-    def state_c_void_p(self):
-        return ctypes.cast(ctypes.pointer(self.val), ctypes.c_void_p)
-
-    @property
-    def size(self):
-        return _DEVICE_POINTER_SIZE
+        value_type = ntype
+        self._cvalue = ctypes.c_void_p(ptr)
+        numba_type = numba.types.CPointer(numba.types.CPointer(value_type))
+        abi_name = f"{self.__class__.__name__}_{str(value_type)}"
+        super().__init__(
+            numba_type=numba_type,
+            value_type=value_type,
+            abi_name=abi_name,
+        )
+
+    @staticmethod
+    def advance(it, distance):
```

Comment:
What do you think about `state` instead of `it` here?

Or `data`, because that's what we're passing in the calls.

Or `it_data` maybe.

(`it` looks too much like it's an actual iterator object.)

Reply:
Agree - I changed it to `state`.